gusto/form_manipulation_labelling.py

```diff
@@ -0,0 +1,328 @@
+    :class:`.linearisation` on t. Returns a labelled form with terms that
+    have the linear form from l=t.get("linearisation") and inherit labels
+    from both t and l, with the labels from l overwriting those from t if
+    both present.
+    """
+    t_linear = t.get(linearisation)
+    assert type(t_linear) == LabelledForm
+    new_labels = t.labels.copy()
+    new_labels["linearisation"] = True
+    return LabelledForm(functools.reduce(
+        operator.add,
+        [Term(l.form, dict(new_labels, **l.labels))
+         for l in t_linear]))
+
+
+def replace_test(new_test):
+    """
+    :arg new_test: a :func:`TestFunction`
+
+    Returns a function that takes in t, a :class:`Term`, and returns
+    a new :class:`Term` with form containing the new_test and
+    labels=t.labels
+    """
+
+    def rep(t):
+        test = t.form.arguments()[0]
+        new_form = ufl.replace(t.form, {test: new_test})
+        return Term(new_form, t.labels)
+
+    return rep
+
+
+def replace_labelled(replacer_in, *labels):
+    """
+    :arg labels: :class:`Label` objects specifying the label(s) of the
+    part of the form that is to be replaced
+    :arg replacer: :class:`ufl.core.expr.Expr`, :func:`Function`,
+    :func:`TrialFunction`, or tuple thereof, or a :class:`Label` object.
+
+    Returns a function that takes in t, a :class:`Term`, and returns
+    a new :class:`Term` where the part of the form labelled by label
+    has been replaced by (possibly part of) the replacer. The labels
+    remain the same, unless the replacer is another labelled part of
+    the form, in which case the original label is removed.
+    """
+
+    def rep(t):
+
+        # if replacer_in is aa label, we need to grab that part of the
+        # form and flag that the original label needs to be removed
+        if type(replacer_in) is Label:
+            replacer = t.get(replacer_in)
+            remove_labels = True
+        else:
+            replacer = replacer_in
+            remove_labels = False
+
+        # dictionary to pass to ufl.replace
+        replace_dict = {}
+
+        for label in labels:
+            # get the thing we're going to replace
+            labelled = t.get(label)
+
+            # return a new term with form and labels identical to t if
+            # no part of the form is labelled with label
+            if labelled is None:
+                return Term(t.form, t.labels)
+
+            # if t.get(label) returns True then the label does not
+            # apply to part of the form so we cannot replace it.
+            if type(labelled) is bool:
+                raise ValueError("This label does not label a part of this term's form")
+
+            # flag if replacer is a ufl expression - if it is then we
+            # trust it is of the right shape below
+            repl_expr = (
+                isinstance(replacer, ufl.core.expr.Expr)
+                and not isinstance(replacer, Function)
+            )
```

Comment:
If we know the shape that it _should_ be, would it be possible to add a check here? That way, we know when we are doing some illegal stuff. I think we should be able to ask for the `ufl_shape` and use that to determine if it's sensible

Reply:
yes, that's probably wise